parser-typechecker/src/U/Codebase/Branch/Diff.hs

```diff
@@ -113,34 +114,34 @@ instance Semigroup NameBasedDiff where
     NameBasedDiff (terms0 <> terms1) (types0 <> types1)
 
 -- | Diff two Branches, returning a tree containing all of the changes
-diffBranches :: forall m. (Monad m) => Branch m -> Branch m -> m TreeDiff
-diffBranches from to = do
+diffBranches :: forall m. (Monad m) => Branch m -> Branch m -> TreeDiff m
+diffBranches from to =
   let termDiffs = diffMap (Branch.terms from) (Branch.terms to)
-  let typeDiffs = diffMap (Branch.types from) (Branch.types to)
-  let defDiff = DefinitionDiffs {termDiffs, typeDiffs}
-  childDiff <- do
-    Align.align (children from) (children to)
-      & wither \case
-        This ca -> do
-          -- TODO: For the names index we really don't need to know which exact
-          -- names were removed, we just need to delete from the index using a
-          -- prefix query, this would be faster than crawling to get all the deletes.
-          removedChildBranch <- Causal.value ca
-          Just . unTreeDiff <$> diffBranches removedChildBranch Branch.empty
-        That ca -> do
-          newChildBranch <- Causal.value ca
-          Just . unTreeDiff <$> diffBranches Branch.empty newChildBranch
-        These fromC toC
-          | Causal.valueHash fromC == Causal.valueHash toC -> do
-              -- This child didn't change.
-              pure Nothing
-          | otherwise -> do
-              fromChildBranch <- Causal.value fromC
-              toChildBranch <- Causal.value toC
-              diffBranches fromChildBranch toChildBranch >>= \case
-                Lens.Empty -> pure Nothing
-                TreeDiff cfr -> pure . Just $ cfr
-  pure $ TreeDiff (defDiff :< childDiff)
+      typeDiffs = diffMap (Branch.types from) (Branch.types to)
+      defDiff = DefinitionDiffs {termDiffs, typeDiffs}
+      childDiff = do
+        Align.align (children from) (children to)
+          & wither \case
+            This ca -> do
+              -- TODO: For the names index we really don't need to know which exact
+              -- names were removed, we just need to delete from the index using a
+              -- prefix query, this would be faster than crawling to get all the deletes.
+              removedChildBranch <- Causal.value ca
+              pure . Just . unTreeDiff $ diffBranches removedChildBranch Branch.empty
+            That ca -> do
+              newChildBranch <- Causal.value ca
+              pure . Just . unTreeDiff $ diffBranches Branch.empty newChildBranch
+            These fromC toC
+              | Causal.valueHash fromC == Causal.valueHash toC -> do
+                  -- This child didn't change.
+                  pure Nothing
```

Comment:
Doesn't need to hold up the pr, but since determining if the key should be included doesn't require doing anything in `m`, couldn't we use `mapMaybe` rather than `wither`? Then `TreeDiff` could be `Cofree (Compose (Map NameSegment) m)` rather than `Cofree (Compose m (Map NameSegment))`. This would allow us to  load one namespace at a time rather than one level at a time. So, we would only run up to two `Causal.value`s at a time rather than running up to 2 * number of subtrees at this level with disagreeing hashes.

Reply:
Yeah this is what I was attempting to do with my original flawed approach 😓 , but forgot to add it back now that things are moved around.

Just implemented it and seems better 😄